src/main/conversions.c

```diff
@@ -1390,118 +1390,69 @@ as_status as_record_init_from_pyobject(AerospikeClient *self, as_error *err,
-                            if (PyErr_ExceptionMatches(PyExc_OverflowError)) {
-                                as_error_update(
-                                    err, AEROSPIKE_ERR_PARAM,
-                                    "integer value exceeds sys.maxsize");
-                            }
-                        }
-                    }
-                    else {
-                        as_error_update(err, AEROSPIKE_ERR_PARAM,
-                                        "TTL should be an int or long");
-                    }
-                }
-                else {
-                    rec->ttl = AS_RECORD_CLIENT_DEFAULT_TTL;
-                }
+        name = PyUnicode_AsUTF8(py_bin_name);
+        if (!name) {
+            as_error_update(err, AEROSPIKE_ERR_CLIENT,
+                            "Unable to convert unicode object to C string");
+            goto CLEANUP;
+        }
 
-                if (py_gen) {
-                    if (PyLong_Check(py_gen)) {
-                        // TODO: need to check that this value does not exceed an unsigned 16 bit integer
-                        rec->gen = (uint16_t)PyLong_AsLong(py_gen);
-                        if (rec->gen == (uint16_t)-1 && PyErr_Occurred()) {
-                            if (PyErr_ExceptionMatches(PyExc_OverflowError)) {
-                                as_error_update(
-                                    err, AEROSPIKE_ERR_PARAM,
-                                    "integer value exceeds sys.maxsize");
-                            }
-                        }
-                    }
-                    else {
-                        as_error_update(err, AEROSPIKE_ERR_PARAM,
-                                        "Generation should be an int or long");
-                    }
-                }
+        if (self->strict_types) {
+            if (strlen(name) > AS_BIN_NAME_MAX_LEN) {
+                as_error_update(
+                    err, AEROSPIKE_ERR_BIN_NAME,
+                    "A bin name should not exceed 15 characters limit");
+                goto CLEANUP;
             }
         }
-        else {
-            rec->ttl = AS_RECORD_CLIENT_DEFAULT_TTL;
+
+        if (!py_bin_value) {
+            // this should never happen, but if it did...
+            as_error_update(err, AEROSPIKE_ERR_CLIENT, "record is null");
+            goto CLEANUP;
         }
 
+        as_val *val = NULL;
+        as_val_new_from_pyobject(self, err, py_bin_value, &val, static_pool,
+                                 serializer_type);
         if (err->code != AEROSPIKE_OK) {
-            as_record_destroy(rec);
+            goto CLEANUP;
+        }
+
+        bool success = as_record_set(rec, name, (as_bin_value *)val);
+        if (success == false) {
+            as_val_destroy(val);
+            as_error_update(err, AEROSPIKE_ERR_BIN_NAME,
+                            "Unable to set key-value pair");
+            goto CLEANUP;
```

Comment:
You don't need to call CLEANUP here.  The record set was not successful and you are already destroying the value.
In the current code, the value of `as_error_update` is returned and `as_record_destroy` is not called.
```
            bool success = as_record_set(rec, name, (as_bin_value *)val);
            if (success == false) {
                as_val_destroy(val);
                return as_error_update(err, AEROSPIKE_ERR_BIN_NAME,
                                       "Unable to set key-value pair");
            }
```
Calling `as_record_destroy` is not free, and this could have a (very slight) impact on performance. I would revert this code to avoid calling `as_record_destroy` unnecessarily.